proposals/2020-06-26_decrypt-max-header-size-max-body-size/proposal.md

```diff
@@ -0,0 +1,275 @@
+
+### Should there be a lower limit for Max Header Size?
+
+To allow for the smallest possible message header
+(no AAD, one EDK but the EDK fields themselves are empty),
+the `Max Header Size` MUST NOT be less than 40 bytes.
+Because future message format revisions might have different lower limits,
+we cannot know if a limit is too low until we start processing a message.
+
+### Will this value still be valid if/when we update the message format?
+
+This control does not introduce any one-way doors that would limit
+how we might update the message format in the future.
+
+Conceptually, this should apply nicely to any new message format.
+With a new message format, our concerns about memory limits should be similar.
+There is not a way around needing to process a whole frame before releasing any plaintext,
+meaning that we will always want some control on the encrypted content size.
+The message header will always contain some metadata that the ESDK might need to hold in memory,
+meaning that a control on the header size would still be useful.
+
+## Reference-level Explanation
+
+Specify two optional inputs on Decrypt, `Max Header Size` and `Max Body Size`.
+
+### Max Body Size
+
+If `max body size` is set on Decrypt, then the operation:
+
+- MUST halt and fail if processing an unframed message with encrypted content field
+  of a length greater than (>) Max Body Size, and MUST NOT attempt decryption on such a message.
+  - This means that when decrypting an unframed message,
+    if Max Body Size is set then the ESDK MUST fail if the encrypted content length field
+    in the message body is greater than (>) Max Body Size.
+- MUST halt and fail if processing a message frame with an encrypted content field
+  of a length greater than (>) Max Body Size, and MUST NOT attempt decryption on any such frame.
+  - This means that when decrypting a regular frame in a framed message,
+    if Max Body Size is set then the ESDK MUST fail if the frame length
+    set in the header is greater than Max Body Size.
+  - This means that when decrypting a final frame in a framed message,
+    if Max Body Size is set then the ESDK MUST fail if the encrypted content length field
+    set in the final frame is greater than Max Body Size.
+  - Note that there is an edge case for messages with bodies that only consist of a final frame.
+    Specifically, a message MAY have a frame size (stated in the header)
+    greater than (>) Max Body Size but the message only contains one final frame,
+    and that final frame has an encrypted content length less than or equal to (<=) Max Body Size.
+    This means that implementations MUST NOT fail
+    based on the frame size in the message header alone,
+    and MUST check the first frame in the message.
+    We allow this is because the Frame Size input is independent of Max Body Size.
+    Frame Size expresses a user intent on encrypt that should not be conflated with
+    the user intent for Max Body Size on decrypt.
```

Comment:
```suggestion
  - Note that there is an edge case for messages with bodies that only consist of a final frame.
    If a message body contains a single frame,
    that final frame MAY have a content length less then (<) the frame size in the header.
    For this reason, the implementation MUST NOT fail
    based on the frame size in the message header alone
    and MUST check the first frame in the message.
    Such a message is valid because the Frame Size is set on encrypt
    while Max Body Size is set only on decrypt.
```